### PR TITLE
Move xdem spatial_tools functionality to geoutils

### PR DESCRIPTION
**Pull request by atedstone**
I hope I've done this properly....

Still to do: delete spatial_tools from xdem and then update all xdem calls to these functions to reference geoutils instead. (Requires activity on the xdem repo)

### CODE REVIEW

**Comment by adehecq** [comment]
There might be some unintended consequences of copy/pasting spatial_tools.py into geoutils:
- some variable naming: dem, ddem... Search for those and replace?
- some functions might not work with multiband rasters

File: geoutils/georaster/spatial_tools.py [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+        transform=rio.transform.from_bounds(*dst_bounds, width=data[0].shape[1], height=data[0].shape[0]),
+        crs=reference_raster.crs,
+        nodata=reference_raster.nodata,
+    )
+
+    return r
+
+
+def merge_rasters(
+    rasters: list[gu.georaster.Raster],
```
Comment (adehecq) [inline geoutils/georaster/spatial_tools.py:218]: Same comments as for stack_rasters overall: make sure it works for any child class.

File: geoutils/georaster/spatial_tools.py [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+        try:
+            merged_data.append(algo(raster_stack.data, axis=0))
+        # If that doesn't work, use the slower np.apply_along_axis approach.
+        except TypeError as exception:
+            if "'axis' is an invalid keyword" not in str(exception):
+                raise exception
+            merged_data.append(np.apply_along_axis(algo, axis=0, arr=raster_stack.data))
+
+    # Save as gu.Raster
+    merged_raster = gu.georaster.Raster.from_array(
```
Comment (adehecq) [inline geoutils/georaster/spatial_tools.py:280]: Use reference_raster.from_array

File: geoutils/georaster/__init__.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
-from geoutils.georaster.raster import Raster, RasterType  # noqa
+from geoutils.georaster.raster import Raster, RasterType  # noqa isort:skip
+from geoutils.georaster import spatial_tools  # noqa
```
[the remark this answers is not in the document]
Comment (atedstone) [reply geoutils/georaster/__init__.py]: moved to directly under geoutils

File: geoutils/georaster/spatial_tools.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+    # Make sure that extent is a multiple of resolution
+    for key1, key2 in zip(("left", "bottom"), ("right", "top")):
+        modulo = (max_bounds[key2] - max_bounds[key1]) % resolution
+        max_bounds[key2] += modulo
+
+    return rio.coords.BoundingBox(**max_bounds)
+
+
+def stack_rasters(
+    rasters: list[gu.georaster.Raster],
```
[the remark this answers is not in the document]
Comment (atedstone) [reply geoutils/georaster/spatial_tools.py]: done

File: geoutils/georaster/spatial_tools.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+    # If no match was found, raise an error.
+    else:
+        raise ValueError(
+            f"'{method_str}' is not a valid rasterio.warp.Resampling method. "
+            f"Valid methods: {[str(method).replace('Resampling.', '') for method in rio.warp.Resampling]}"
+        )
+    return resampling_method
+
+
+def merge_bounding_boxes(bounds: list[rio.coords.BoundingBox], resolution: float) -> rio.coords.BoundingBox:
```
[the remark this answers is not in the document]
Comment (atedstone) [reply geoutils/georaster/spatial_tools.py]: I'm not familiar with either of these functions - perhaps this is something that you could have a go at? (or raise another issue to be dealt with in the future)

[the remark this answers is not in the document]
Comment (adehecq) [reply geoutils/georaster/spatial_tools.py]: Just raised the issue.

**Pull request merged**